cmd/ttnv3/ttnv3.go

```diff
@@ -0,0 +1,42 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttnv3
+
+import (
+	"github.com/spf13/cobra"
+	"go.thethings.network/lorawan-stack-migrate/pkg/source"
+)
+
+const sourceName = "ttnv3"
+
+// TTNv3Cmd represents the ttnv3 source
+var TTNv3Cmd = &cobra.Command{
+	Use:          sourceName + " ...",
+	Short:        "Export devices from The Things Stack",
+	SilenceUsage: true,
+	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+		// Execute root PersistentPreRunE
+		root := *cmd
+		for root.HasParent() { // Find absolute root
+			root = *root.Parent()
+		}
+		if err := root.PersistentPreRunE(cmd, args); err != nil {
+			return err
+		}
+
+		source.RootConfig.Source = sourceName
+		return nil
+	},
```

Comment:
@adriansmares This doesn't seem ideal. If `PersistentPreRunE` is declared in subcommand it overwrites parent's persistent pre run. In `rootCmd`'s pre run we have some initialisation code that has to run for every execution.
This pre run would be used to set the source for all it's subcommands to `ttnv3`.
Do you have any better ideas on implementing this without a major design change?

Reply:
I think this will work, and you probably can get the logic out of this package somehow. 

But are you sure this is correct ? Right now if `root` does not have a parent, you're recursing indefinitely, right ? Also I don't fully see why you need the pointer dereference.

Just execute `PersistentPreRunE` if `cmd.HasParent()` and `cmd.Parent().PersistentPreRunE` is not `nil`.

Reply:
> But are you sure this is correct ? Right now if root does not have a parent, you're recursing indefinitely, right ?

If root has no parent the statement will be false and the loop won't execute anymore, so it finds the absolute root (`rootCmd`).

> Just execute PersistentPreRunE if cmd.HasParent() and cmd.Parent().PersistentPreRunE is not nil.

We can do this with the caveat that each subcommand that doesn't have `Persistent{Pre,Post}Run{,e}` declared inherits it from parent, in which case parent's pre-run would be executed more than once (e.g. in `applicationsCmd` and in `TTNv3Cmd`). 
We could have a function in `pkg/commands`
```go
func ExecuteParentPersistentPreRun(cmd *cobra.Command, args []string) error {
	if !cmd.HasParent() {
		return nil
	}
	p := cmd.Parent()

	if f := p.PersistentPreRunE; f != nil {
		if err := f(p, args); err != nil {
			return err
		}
	} else if f := p.PersistentPreRun; f != nil {
		f(p, args)
	}
	return nil
}
```
and add `PersistentPreRunE: commands.ExecuteParentPersistentPreRun` if we don't want this to happen.

The persistent pre-run of `TTNv3Cmd` would become
```suggestion
	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
		source.RootConfig.Source = sourceName

		return commands.ExecuteParentPersistentPreRun(cmd, args)
	},
```

Reply:
> I mean that if cmd does not have a parent, root == cmd and then when you do root.PersistentPreRunE you end up calling the same function, thus recursing, right ?

I get it now. Yes, but I assumed the only cmd without a parent is the `rootCmd` for this draft.

I'll go with the second approach.